apps/indexer/src/services/events/event-outbox-fan-out-service.ts

```diff
@@ -0,0 +1,142 @@
+import { sql } from "drizzle-orm";
+import Deferred from "promise-deferred";
+import { schema } from "../../../schema.ts";
+import type { DB } from "../db.ts";
+
+type EventOutboxFanOutService_FanOutEventsParams = {
+  signal: AbortSignal;
+};
+
+type EventOutboxFanOutServiceOptions = {
+  db: DB;
+  /**
+   * Polling interval in milliseconds for checking for new events.
+   * This works as fallback to LISTEN/NOTIFY if there are no new events.
+   *
+   * @default 30_000
+   */
+  pollInterval?: number;
+};
+
+export class EventOutboxFanOutService {
+  private readonly db: DB;
+  private readonly pollInterval: number;
+  private shouldProcessBatch: boolean;
+  private deferred: Deferred.Deferred<void> | undefined;
+
+  constructor(options: EventOutboxFanOutServiceOptions) {
+    this.db = options.db;
+    this.pollInterval = options.pollInterval ?? 30_000;
+    this.shouldProcessBatch = false;
+  }
+
+  async fanOutEvents(
+    params: EventOutboxFanOutService_FanOutEventsParams,
+  ): Promise<void> {
+    const { signal } = params;
+
+    const notificationPgClient = await this.db.$client.connect();
+
+    if (!notificationPgClient) {
+      throw new Error("Failed to connect to notification pg client");
+    }
+
+    const processBatchListener = () => {
+      // if there is a deferred, resolve it
+      if (this.deferred) {
+        this.deferred.resolve();
+      } else {
+        // this will be picked by the waitForNewDeliveries method
+        // and immediately resolve the new promise
+        this.shouldProcessBatch = true;
+      }
+    };
+
+    // start listening for new events using notifications
+    await notificationPgClient.query("LISTEN event_outbox_events");
+    notificationPgClient.on("notification", processBatchListener);
+
+    // fallback to polling if there were no notifications for {this.interval} seconds
+    const intervalId = setInterval(processBatchListener, this.pollInterval);
+
+    // clean up because service is aborted
+    signal.addEventListener("abort", () => {
```

Comment:
since there were one `await` above this line. the `signal` can be aborted before this line is executed, when that happens, the `notificationPgClient` and event listener will leak.